src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Pkcs12CertBag.cs

```diff
@@ -0,0 +1,125 @@
+        /// </summary>
+        /// <param name="certificateType">The identifier for the certificate type</param>
+        /// <param name="encodedCertificate">The encoded value</param>
+        /// <remarks>
+        /// No validation is done to ensure that the <paramref name="encodedCertificate"/> value is
+        /// correct for the indicated <paramref name="certificateType"/>.  Note that for X.509
+        /// public-key certificates the correct encoding for a CertBag value is to wrap the
+        /// DER-encoded certificate in an OCTET STRING.
+        /// </remarks>
+        public Pkcs12CertBag(Oid certificateType, ReadOnlyMemory<byte> encodedCertificate)
+            : base(
+                Oids.Pkcs12CertBag,
+                EncodeBagValue(certificateType, encodedCertificate),
+                skipCopy: true)
+        {
+            _certTypeOid = new Oid(certificateType);
+
+            _decoded = AsnSerializer.Deserialize<CertBagAsn>(
+                EncodedBagValue,
+                AsnEncodingRules.BER);
+
+            IsX509Certificate = _decoded.CertId == Oids.Pkcs12X509CertBagType;
+        }
+
+        internal Pkcs12CertBag(X509Certificate2 cert)
+            : base(
+                Oids.Pkcs12CertBag,
+                EncodeBagValue(
+                    Oids.Pkcs12X509CertBagType,
+                    PkcsPal.Instance.EncodeOctetString(cert.RawData)),
+                skipCopy: true)
+        {
+            _decoded = AsnSerializer.Deserialize<CertBagAsn>(
+                EncodedBagValue,
+                AsnEncodingRules.BER);
+
+            IsX509Certificate = true;
+        }
+
+        public Oid GetCertificateType()
+        {
+            if (_certTypeOid == null)
+            {
+                _certTypeOid = new Oid(_decoded.CertId);
+            }
+
+            return new Oid(_certTypeOid);
+        }
+
+        public ReadOnlyMemory<byte> EncodedCertificate => _decoded.CertValue;
+
+        public X509Certificate2 GetCertificate()
+        {
+            if (!IsX509Certificate)
+            {
+                throw new InvalidOperationException(SR.Cryptography_Pkcs12_CertBagNotX509);
+            }
+
+            return new X509Certificate2(Helpers.DecodeOctetString(_decoded.CertValue).ToArray());
+        }
+
+        private static byte[] EncodeBagValue(Oid certificateType, ReadOnlyMemory<byte> encodedCertificate)
+        {
+            if (certificateType == null)
+                throw new ArgumentNullException(nameof(certificateType));
+
+            return EncodeBagValue(certificateType.Value, encodedCertificate);
+        }
+
+        private static byte[] EncodeBagValue(string certificateType, ReadOnlyMemory<byte> encodedCertificate)
+        {
+            // Read to ensure that there is precisely one legally encoded value.
+            AsnReader reader = new AsnReader(encodedCertificate, AsnEncodingRules.BER);
+            reader.GetEncodedValue();
+            reader.ThrowIfNotEmpty();
```

Comment:
The comment says "precisely one".  Naively from these method names I'd assume this was verifying >= 1 rather than == 1.

Reply:
`GetEncodedValue` throws if there's not a value, then consumes the value.  `ThrowIfNotEmpty` throws if there is still data.

Reply:
Ah, I misread ThrowIfNotEmpty as ThrowIfEmpty.